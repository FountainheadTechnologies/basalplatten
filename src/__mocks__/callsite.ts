module.exports = () => ([{
}, {
  getFileName: () => 'test-caller.js'
}]);
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
