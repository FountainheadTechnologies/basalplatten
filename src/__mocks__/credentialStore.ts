export const credentialStore = {
  setToken: jest.fn(),
  getToken: jest.fn(),
  clearToken: jest.fn()
}
