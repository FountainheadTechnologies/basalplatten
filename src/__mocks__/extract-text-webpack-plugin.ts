module.exports = class ExtractTextPlugin {
  extract() {}
};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
