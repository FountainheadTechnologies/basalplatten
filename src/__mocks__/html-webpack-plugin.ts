module.exports = class HtmlWebpackPlugin {};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
