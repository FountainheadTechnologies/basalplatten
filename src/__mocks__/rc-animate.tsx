import React from 'react';

const RcAnimate = props => (
  <div>{props.component}</div>
)

RcAnimate.displayName = 'RcAnimate';

export default RcAnimate;
