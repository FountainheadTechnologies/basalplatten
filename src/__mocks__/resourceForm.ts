const submit = jest.fn();

export const createToken = {
  schema: {
    required: ['username', 'password'],
    properties: {
      username: { type: 'string' },
      password: { type: 'string' }
    },
  },

  submit
}

export const createPerson = {
  schema: {
    required: ['forename', 'surname'],
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  },

  submit
}

export const updatePerson = {
  schema: {
    required: [],
    default: {
      forename: 'Test',
      surname: 'Person',
      date_of_birth: '1985-08-02T23:00:00.000Z',
      credits: 33,
      favourite_colour: 'blue'
    },
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      credits: { type: 'number' },
      enabled: { type: 'boolean' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      },
      favourite_colour: {
        enum: ['red', 'green', 'blue']
      }
    }
  }
}
