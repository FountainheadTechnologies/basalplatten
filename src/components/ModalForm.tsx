import React from 'react';
import { FormItemProps } from 'antd/lib/form/FormItem';

import { buttonizedModal, InjectedProps } from '../hoc/buttonizedModal';
import { Form, FormComponent, Props as FormProps } from './Form';

// @hack: Required to work around 'cannot be named' TS bug
import { ButtonProps } from 'antd/lib/button/button';
import { ModalProps } from 'antd/lib/modal/Modal';
import { State as BMState, Props as BMProps } from '../hoc/buttonizedModal';
export { ButtonProps, ModalProps, BMState, BMProps };

export interface Props {
  form: FormProps & {
    children?: React.ReactChildren
  };
  onSuccess?: (value: any) => void;
  onFailure?: (error: any) => void;
}

const DEFAULT_ITEM_PROPS: Partial<FormItemProps> = {
  labelCol: {
    span: 8
  },

  wrapperCol: {
    span: 16
  }
}

export class ModalFormComponent extends React.Component<Props & InjectedProps> {
  protected _form: FormComponent;

  render() {
    return (
      <Form
        defaultItemProps={DEFAULT_ITEM_PROPS}
        {...this.props.form}
        submitButton={false}
        onSuccess={this.onSuccess}
        wrappedComponentRef={(form: FormComponent) => this._form = form}
      >
        {this.props.form.children || this.props.children}
      </Form>
    );
  }

  componentWillMount() {
    this.props.setOkHandler(this._submit);
  }

  protected _submit = () => {
    this.props.setWorking(true);

    return this._form
      .submit()
      .finally(() => {
        this.props.setWorking(false);
      });
  }

  onSuccess = () => {
    this.props.setVisible(false);
  }
}

export const ModalForm = buttonizedModal<Props & BMProps>(ModalFormComponent, props => ({
  modal: {
    okText: 'Save',
    cancelText: 'Cancel'
  }
}));
