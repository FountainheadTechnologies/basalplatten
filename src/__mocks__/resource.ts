import { createToken } from './resourceForm';

export const api = {
  properties: {},
  meta: {},

  config: {
    headers: {}
  },

  formNamed(rel: string, name: string) {
    if (rel === 'token' && name === 'create') {
      return createToken;
    }
  }
}
