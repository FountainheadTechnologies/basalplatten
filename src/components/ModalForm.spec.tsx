import React from 'react';
import promiseFinally from 'promise.prototype.finally';
import { shallow } from 'enzyme';

import { ModalForm, ModalFormComponent } from './ModalForm';
import { createPerson } from '../__mocks__/resourceForm';

promiseFinally.shim();

describe('default behaviour', () => {
  const root = shallow(
    <ModalForm />
  );

  it('renders a button', () => {
    expect(root).toMatchSnapshot();
  });
});

describe('wrapped component behaviour', () => {
  let _okHandler;
  const onSuccess = jest.fn();
  const setWorking = jest.fn();
  const setVisible = jest.fn();
  const setOkHandler = jest.fn(handler => {
    _okHandler = handler;
  });

  const root = shallow(
    <ModalFormComponent
      form={{ resource: createPerson }}
      setWorking={setWorking}
      setVisible={setVisible}
      setOkHandler={setOkHandler}
      onSuccess={onSuccess}
    />
  );

  it('calls `props.setWorking` while form is submitting', () => {
    let _p;
    let _resolve;

    root.instance()._form = {
      submit: () => {
        _p = new Promise((resolve) => {
          _resolve = resolve;
        })

        return _p;
      }
    };

    const handler = _okHandler().then(() => {
      expect(setWorking).toHaveBeenCalledWith(false);
    });

    expect(setWorking).toHaveBeenCalledWith(true);
    _resolve();

    return handler;
  });

  it('hides the form when submission succeeded', () => {
    root.props().onSuccess();
    expect(setVisible).toHaveBeenCalledWith(false);
    expect(onSuccess).toHaveBeenCalled();
  });
});
