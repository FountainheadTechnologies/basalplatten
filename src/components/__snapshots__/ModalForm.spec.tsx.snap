// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behaviour renders a button 1`] = `
<Button
  clicked={false}
  ghost={false}
  loading={false}
  onClick={[Function]}
  prefixCls="ant-btn"
/>
`;
